small/ibuf.c

```diff
@@ -104,3 +104,41 @@ ibuf_reserve_slow(struct ibuf *ibuf, size_t size)
 	return ibuf->wpos;
 }
 
+void
+ibuf_shrink(struct ibuf *ibuf)
+{
+	size_t used = ibuf_used(ibuf);
+	if (used == 0) {
+		if (ibuf->buf)
+			slab_put(ibuf->slabc, slab_from_data(ibuf->buf));
+		ibuf->buf = ibuf->rpos = ibuf->wpos = ibuf->end = NULL;
+		return;
+	}
+
+	size_t capacity = ibuf_capacity(ibuf);
+	size_t new_capacity = ibuf->start_capacity;
+	if (new_capacity < used)
+		new_capacity = used;
+	if (capacity == new_capacity)
+		return;
+
+	/* Avoid relocation if the actual slab size doesn't change. */
+	uint8_t order, new_order;
+	order = slab_order(ibuf->slabc, capacity + slab_sizeof());
+	new_order = slab_order(ibuf->slabc, new_capacity + slab_sizeof());
+	if (order == new_order && order <= ibuf->slabc->order_max)
```

Comment:
Nit: Too many implicit assumptions in this code IMO:
 - The actual size of a slab is less than the allocated size by `slab_sizeof()`.
 - If `slab_order()` equals `order_max + 1`, then the slab is allocated from `malloc`.

This is probably fine as long as this code lives in the same library as the slab arena implementation. Still, I'd prefer to add a helper function that would compute the actual size of a slab returned by `slab_get()`. Up to @alyapunov.

Reply:
Done.